### PR TITLE
Convert `makeCredential()`'s parameters into a dictionary.

### PR DESCRIPTION
**Pull request by mikewest**
Passing a single dictionary parameter into `makrCredential()` provides
for greater forward compatibility, as new data can be flexibly added
to the method invocation without restructuring the existing
structure. It also helps developers understand what they're passing
in, as each parameter will be labeled.

This patch restructures the data passed into `makeCredential()`
substantially, moving from four parameters to a single dictionary,
and merging some existing types into a simpler structure. Most of
it is straightforward; the only bit I know will be controversial is
dropping `RelyingPartyUserInfo` in favor of two instances of a
simpler `ScopedCredentialEntity` object: one for the RP, one for the
user.

Let's chat about how (un)reasonable this approach might be.

### CODE REVIEW

**Comment by mikewest** [comment]
As @vijaybh suggeted in #2 of https://lists.w3.org/Archives/Public/public-webauthn/2017Apr/0158.html, moving getAssertion() and makeCredential() to dictionary-based invocation rather than list-of-parameter-based invocation seems pretty reasonable. This patch migrates `makeCredential()`, and makes some potentially controversial suggestions in doing so. I've kept it separate from the `getAssertion()` migration for that reason.

File: index.bs [hunk 1]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
     [[#op-get-assertion]].
 </div>
 
+## Additional options for Attestation (dictionary <dfn dictionary>MakeCredentialOptions</dfn>) ## {#dictionary-makecredentialoptions}
```
Comment (vijaybh) [inline index.bs]: Editorial: this isn't "additional options" any more, it's all the options. Maybe say something like "Parameters for credential creation" instead?

Comment (mikewest) [inline index.bs]: Picked up in https://github.com/w3c/webauthn/pull/399/commits/6dca154104884ed814f2b0ded74a959bdf28e3ef.

Comment (equalsJeffH) [inline index.bs]: yeah, as i noted in review of PR #398 maybe these are not "options" but are parameters, some of which are optional ?

File: index.bs [hunk 2]
```diff
@@ -377,12 +377,7 @@ The Web Authentication API is defined by the union of the Web IDL fragments pres
 <xmp class="idl">
     [SecureContext]
     interface WebAuthentication {
-        Promise<ScopedCredentialInfo> makeCredential(
-            RelyingPartyUserInfo                 accountInformation,
-            sequence<ScopedCredentialParameters> cryptoParameters,
-            BufferSource                         attestationChallenge,
-            optional ScopedCredentialOptions     options
-        );
+        Promise<ScopedCredentialInfo> makeCredential(MakeCredentialOptions options);
```
Comment (leshi) [inline index.bs]: \<bikeshed> I like the `request` terminology that you had in PR #398.  Maybe also rename `MakeCredentialOptions` to `MakeRequest` or `CredentialRequest`? \</bikeshed>

File: index.bs [hunk 3]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
...
+
+<xmp class="idl">
+    dictionary MakeCredentialOptions {
+        // Relying Party / User Metadata:
+        required ScopedCredentialEntity rp;
+        required ScopedCredentialEntity user;
+
+        // Attestation configuration:
+        required BufferSource                         challenge;
+        required sequence<ScopedCredentialParameters> parameters;
```
Comment (leshi) [inline index.bs]: I don't think these are the attestation parameters -- they're the parameters for the type of credential.  We don't have parameters that allow an RP to request a particular type of attestation, though maybe we should.

File: index.bs [hunk 4]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
     [[#op-get-assertion]].
 </div>
 
+## Parameters for Credential Creation (dictionary <dfn dictionary>MakeCredentialOptions</dfn>) ## {#dictionary-makecredentialoptions}
+
+<xmp class="idl">
+    dictionary MakeCredentialOptions {
+        // Relying Party / User Metadata:
+        required ScopedCredentialEntity rp;
```
Comment (leshi) [inline index.bs]: Interesting, so we now have the same data for an rp and the user.  Cool!

File: index.bs [hunk 5]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
...
+        of registered WebAuthn Extensions.
+</div>
+
+### Entity Description ### {#dictionary-scopedcredentialentity}
+
+The {{ScopedCredentialEntity}} dictionary describes a user account or a [=relying party=] with which a credential is associated.
+
+<xmp class="idl">
+    dictionary ScopedCredentialEntity {
+        DOMString id;
```
Comment (leshi) [inline index.bs]: In `RelyingPartyUserInfo` some of these were required, why the change?

File: index.bs (hunk 2 above)
Comment (equalsJeffH) [inline index.bs]: But, I do not like the term `request` here because that dict instance is not a "request" in an of itself.  I much prefer just using `options` or `parameters` here (see https://github.com/w3c/webauthn/pull/398#issuecomment-294038181).  If I were to implement, I'd find `request` misleading/confusing.

File: index.bs [hunk 6]
```diff
@@ -443,19 +417,22 @@ When this method is invoked, the user agent MUST execute the following algorithm
     |callerOrigin| is an [=opaque origin=], return [=a promise rejected with=] a {{DOMException}} whose name is
     "{{NotAllowedError}}", and terminate this algorithm.
 
-1. If the {{ScopedCredentialOptions/rpId}} member of {{options}} is [=present|not present=], then set |rpId| to |callerOrigin|.
-    Otherwise:
+1. If the {{ScopedCredentialEntity/id}} member of {{options}}.{{MakeCredentialOptions/rp}} is [=present|not present=], then set
+    |rpId| to |callerOrigin|.
+
+   Otherwise:
```
Comment (equalsJeffH) [inline index.bs]: unfortunately, the as-rendered indent of this "otherwise" block is now lost and so it is visually confusing that this is a block of substeps. we need to fix this.

File: index.bs [hunk 7]
```diff
@@ -443,19 +417,22 @@ When this method is invoked, the user agent MUST execute the following algorithm
...
-        return [=a promise rejected with=] a {{DOMException}} whose name is "{{SecurityError}}", and terminate this algorithm.
-    1. Set |rpId| to the {{ScopedCredentialOptions/rpId}}.
+    1. If {{options}}.{{MakeCredentialOptions/rp}}.{{ScopedCredentialEntity/id}} [=is not a registrable domain suffix of and is
+        not equal to=] |effectiveDomain|, return [=a promise rejected with=] a {{DOMException}} whose name is
+        "{{SecurityError}}", and terminate this algorithm.
+    1. Set |rpId| to {{options}}.{{MakeCredentialOptions/rp}}.{{ScopedCredentialEntity/id}}.
 
-1. Let |normalizedParameters| be a new [=list=] whose [=list/items=] are pairs of ScopedCredentialType and a [=dictionary=] type
-    (as returned by [=normalizing an algorithm=]).
+1. Let |normalizedParameters| be a new [=list=] whose [=list/items=] are pairs of {{ScopedCredentialType}} and a [=dictionary=]
```
Comment (equalsJeffH) [inline index.bs]: unfortunately, now the as-rendered step numbering is re-started here, rather than properly continuing.

File: index.bs (hunk 3 above)
Comment (equalsJeffH) [inline index.bs]: Hm. the challenge is not a part of attestation per-se -- it provides per-protocol-run roundtrip integrity whether the protocol run is a registration operation or an authn operation.  

ScopedCredentialParameters are params for the creation of the underlying credential key pair, which then is attested.

File: index.bs [hunk 8]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
     [[#op-get-assertion]].
 </div>
 
+## Parameters for Credential Creation (dictionary <dfn dictionary>MakeCredentialOptions</dfn>) ## {#dictionary-makecredentialoptions}
```
Comment (equalsJeffH) [inline index.bs]: nominally, I think I'm ok with this overall restructuring of the makeCredential parameters/options, tho I have detail-level comments below.

File: index.bs [hunk 9]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
...
+        required ScopedCredentialEntity rp;
+        required ScopedCredentialEntity user;
+
+        // Attestation configuration:
+        required BufferSource                         challenge;
+        required sequence<ScopedCredentialParameters> parameters;
+
+        // Optional constraints:
+        unsigned long                        timeout;
+        sequence<ScopedCredentialDescriptor> exclude;
```
Comment (equalsJeffH) [inline index.bs]: I would keep the name of this as `excludeList`.  I note that the other PRs do not alter `allowList`'s name.

File: index.bs [hunk 10]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
...
+
+    :   <dfn>user</dfn>
+    ::  This member contains data about the user account for which the [=relying party=] is requesting attestation.
+
+        Its value's {{ScopedCredentialEntity/name}} member is required, and contains a friendly name for the user account (e.g.
+        "john.p.smith@example.com", or "John P. Smith").
+
+        Its value's {{ScopedCredentialEntity/id}} member is required, and contains an identifier for the account, specified by
+        the [=relying party=]. This is not meant to be displayed to the user, but is used by the relying party to control the
+        number of credentials - an authenticator will never contain more than one credential for a given relying party under the
```
Comment (equalsJeffH) [inline index.bs]: aside: see https://github.com/w3c/webauthn/issues/403

File: index.bs [hunk 11]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
...
+    :   <dfn>parameters</dfn>
+    ::  This member contains information about the desired properties of the credential to be created. The sequence is ordered
+        from most preferred to least preferred. The platform makes a best-effort to create the most preferred credential that it
+        can.
+
+    :   <dfn>timeout</dfn>
+    ::  This member specifies a time, in milliseconds, that the caller is willing to wait for the call to complete. This is
+        treated as a hint, and may be overridden by the platform.
+
+    :   <dfn>exclude</dfn>
```
Comment (equalsJeffH) [inline index.bs]: s/exclude/excludeList/

File: index.bs [hunk 12]
```diff
@@ -889,6 +792,102 @@ user consent to a specific transaction. The structure of these signatures is def
...
+        Its value's {{ScopedCredentialEntity/name}} member is required, and contains a friendly name for the user account (e.g.
+        "john.p.smith@example.com", or "John P. Smith").
+
+        Its value's {{ScopedCredentialEntity/id}} member is required, and contains an identifier for the account, specified by
+        the [=relying party=]. This is not meant to be displayed to the user, but is used by the relying party to control the
+        number of credentials - an authenticator will never contain more than one credential for a given relying party under the
+        same {{ScopedCredentialEntity/id}}.
+
+    :   <dfn>challenge</dfn>
+    ::  This member contains a channelge intended to be used for generating the newly created credential's [=attestation object=].
```
Comment (equalsJeffH) [inline index.bs]: s/channelge/challenge/ :)

need to expand the explanation (but not necessarily now) -- challenge is used in both regstn and authn protocol runs and has security implications. have submitted https://github.com/w3c/webauthn/issues/404

File: index.bs (hunk 5 above)
Comment (equalsJeffH) [inline index.bs]: +1 to @leshi's question "In `RelyingPartyUserInfo` some of these were required, why the change?"

File: index.bs (hunk 2 above)
Comment (leshi) [inline index.bs]: I'm fine with whatever term we use, as long as it's consistent

File: index.bs (hunk 5 above)
Comment (leshi) [inline index.bs]: Suggest making `id` and `name` required.

Comment (leshi) [inline index.bs]: On the call, Jeffrey & Mike mentioned that requiring developers to set RP is a bad idea and the params that were required in RelyingPartyUserInfo can be enforced in the algorithm.

File: index.bs (hunk 3 above)
Comment (mikewest) [inline index.bs]: If y'all have suggestions for both `parameters` and `Attestation configuration`, I'm all ears. I don't have strong opinions about either.

File: index.bs (hunk 2 above)
Comment (mikewest) [inline index.bs]: I think we agreed that we're running with `options` for both.

File: index.bs (hunk 3 above)
Comment (mikewest) [inline index.bs]: I think we agreed on the call to drop the comment, and leave the names.

**Comment by mikewest** [comment]
Addressed the comments based on my understanding of what we agreed to in the call via https://github.com/w3c/webauthn/pull/399/commits/805891752a579d84c677f512539c34f05658cf8d, and merged in #398 (and therefore #397) via https://github.com/w3c/webauthn/pull/399/commits/ae33b140cd808f849c9c238d3b703099bc3cd9df. Land those two before you land this one. :)

File: index.bs (hunk 3 above)
Comment (equalsJeffH) [inline index.bs]: i think "drop the comment, and leave the names" is fine for now", thx.

**Comment by equalsJeffH** [comment]
LGTM, thx @mikewest !
